apps/emqx/src/emqx_router_syncer.erl

```diff
@@ -0,0 +1,328 @@
+    %% operations and go down to low priority ones. This might cause some operations to
+    %% stay in stash for unfairly long time, when there are many high priority operations.
+    %% However, it's unclear how likely this is to happen in practice.
+    mk_batch(Stash, ?MAX_BATCH_SIZE).
+
+mk_batch(Stash, BatchSize) ->
+    mk_batch(?PRIO_HI, #{}, BatchSize, Stash).
+
+mk_batch(Prio, Batch, SizeLeft, Stash) ->
+    mk_batch(Prio, Batch, SizeLeft, Stash, maps:iterator(Stash)).
+
+mk_batch(Prio, Batch, SizeLeft, Stash, It) when SizeLeft > 0 ->
+    %% Iterating over stash, only taking operations with priority equal to `Prio`.
+    case maps:next(It) of
+        {Route, Op = ?ROUTEOP(_Action, Prio), NIt} ->
+            NBatch = Batch#{Route => Op},
+            NStash = maps:remove(Route, Stash),
+            mk_batch(Prio, NBatch, SizeLeft - 1, NStash, NIt);
+        {_Route, _Op, NIt} ->
+            %% This is lower priority operation, skip it.
+            mk_batch(Prio, Batch, SizeLeft, Stash, NIt);
+        none ->
+            %% No more operations with priority `Prio`, go to the next priority level.
+            true = Prio < ?PRIO_BG,
+            mk_batch(Prio + 1, Batch, SizeLeft, Stash)
+    end;
+mk_batch(_Prio, Batch, _, Stash, _It) ->
+    {Batch, Stash}.
+
+send_replies(Errors, Batch) ->
+    maps:foreach(
+        fun(Route, {_Action, _Prio, Ctx}) ->
+            case Ctx of
+                [] ->
+                    ok;
+                _ ->
+                    replyctx_send(maps:get(Route, Errors, ok), Ctx)
+            end
+        end,
+        Batch
+    ).
+
+replyctx_send(_Result, []) ->
+    noreply;
+replyctx_send(Result, {MRef, Pid}) ->
+    Pid ! {MRef, Result}.
+
+%%
+
+run_batch(Batch) ->
+    emqx_router:do_batch(Batch).
+
+%%
+
+stash_new() ->
+    #{}.
+
+stash_empty(Stash) ->
+    maps:size(Stash) =:= 0.
+
+stash_drain(Stash) ->
+    receive
+        ?PUSH(Prio, Op) ->
+            stash_drain(stash_add(Prio, Op, Stash))
+    after 0 ->
+        Stash
+    end.
+
+stash_add(Pushes, Stash) ->
+    lists:foldl(
+        fun(?PUSH(Prio, Op), QAcc) -> stash_add(Prio, Op, QAcc) end,
+        Stash,
+        Pushes
+    ).
+
+stash_add(Prio, ?OP(Action, Topic, Dest, Ctx), Stash) ->
+    Route = {Topic, Dest},
+    case maps:get(Route, Stash, undefined) of
+        undefined ->
+            Stash#{Route => {Action, Prio, Ctx}};
+        RouteOp ->
+            case merge_route_op(RouteOp, ?ROUTEOP(Action, Prio, Ctx)) of
+                undefined ->
+                    maps:remove(Route, Stash);
+                RouteOpMerged ->
+                    Stash#{Route := RouteOpMerged}
+            end
+    end.
+
+merge_route_op(?ROUTEOP(Action, _Prio1, Ctx1), DestOp = ?ROUTEOP(Action)) ->
+    %% NOTE: This should not happen anyway.
+    _ = replyctx_send(ignored, Ctx1),
+    DestOp;
+merge_route_op(?ROUTEOP(_Action1, _Prio1, Ctx1), ?ROUTEOP(_Action2, _Prio2, Ctx2)) ->
+    %% NOTE: Operations cancel each other.
```

Comment:
this assumes the base is always consitent.
to be more idempotent, maybe only cancel the old op and keep the new op?

Reply:
Good idea.

> this assumes the base is always consitent.

Worth noting that it's not really "always", e.g. add + delete in _different_ batches won't be able to cancel each other.